relay-sampling/src/lib.rs

```diff
@@ -745,23 +749,52 @@ pub struct DynamicSamplingContext {
     /// Set on transaction start, or via `scope.transaction`.
     #[serde(default)]
     pub transaction: Option<String>,
+    /// The same rate with which this trace was sampled. This is a float between 0 and 1.
+    #[serde(default)]
+    pub sample_rate: Option<JsonStringifiedValue<f64>>,
     /// The user specific identifier ( e.g. a user segment, or similar created by the SDK
     /// from the user object).
     #[serde(flatten, default)]
     pub user: TraceUserContext,
 }
 
 impl DynamicSamplingContext {
+    /// Compute the effective sampling rate based on the random "diceroll" and the sample rate from
+    /// the matching rule.
+    pub fn adjust_sample_rate(&self, rule_sample_rate: f64) -> f64 {
+        let client_sample_rate = self.sample_rate.map(|x| x.0).unwrap_or(1.0);
+        let mut adjusted_sample_rate = (rule_sample_rate / client_sample_rate).clamp(0.0, 1.0);
+
+        if adjusted_sample_rate.is_infinite() || adjusted_sample_rate.is_nan() {
+            // adjusted_sample_rate is infinite or NaN. This is a bug.
+            //
+            // - infinite should not happen because we clamped the number.
+            // - NaN can happen if the client_sample_rate is 0, which is bogus because the SDK
+            //   should've dropped the envelope. In that case let's pretend the sample rate was
+            //   not sent, because clearly the sampling decision across the trace is still 1.
+            //   The most likely explanation is that the SDK is reporting its own sample rate
+            //   setting instead of the one from the continued trace.
+            //
+            // since we write back the client_sample_rate into the event's trace context, it should
+            // be possible to find those values + sdk versions via snuba
+            relay_log::warn!("adjusted sample rate ended up being inf/nan");
```

Comment:
nit: Not sure if we already have enough context on the sentry scope to identify the project or sdk version sending this, but we can also add that when the warning shows up (if it ever will).

Reply:
we don't have enough context here right now, i think it's easy enough to look into snuba